plugins/modules/packaging/language/cargo.py

```diff
@@ -0,0 +1,192 @@
+        self.state = kwargs['state']
+        self.version = kwargs['version']
+
+        self.executable = [module.get_bin_path('cargo', True)]
+
+    @property
+    def installed_packages(self):
+        cmd = ['install', '--list']
+        data = self._exec(cmd, True, False, False)
+
+        package_regex = re.compile(r'^(\w+) v(.+):$')
+        installed_packages = {}
+        for line in data.splitlines():
+            package_info = package_regex.match(line)
+            if package_info:
+                installed_packages[package_info[1]] = package_info[2]
+
+        return installed_packages
+
+    @property
+    def path(self):
+        return self._path
+
+    @path.setter
+    def path(self, path):
+        if path is not None and not os.path.isdir(path):
+            self.module.fail_json(msg="Path %s is not a directory" % path)
+        self._path = path
+
+    def _exec(self, args, run_in_check_mode=False, check_rc=True, add_package_name=True):
+        if not self.module.check_mode or (self.module.check_mode and run_in_check_mode):
+            cmd = self.executable + args
+            rc, out, err = self.module.run_command(cmd, check_rc=check_rc)
+            return out
+        return ''
+
+    def install(self):
+        cmd = ['install']
+        if self.name:
+            cmd.append(self.name)
+        if self.path:
+            cmd.append('--root')
+            cmd.append(self.path)
+        if self.version:
+            cmd.append('--version')
+            cmd.append(self.version)
+        return self._exec(cmd)
+
+    def is_outdated(self, name):
+        installed_version = self.installed_packages.get(name)
+
+        cmd = ['search', name, '--limit', '1']
+        data = self._exec(cmd, True, False, False)
+
+        match = re.search(r'"(.+)"', data)
+        if match:
+            latest_version = match[1]
+
+        return installed_version != latest_version
+
+    def uninstall(self):
+        cmd = ['uninstall']
+        if self.name:
+            cmd.append(self.name)
+        return self._exec(cmd)
+
+
+def main():
+    arg_spec = dict(
+        name=dict(required=True, type='str'),
+        path=dict(default=None, type='path'),
+        state=dict(default='present', choices=['present', 'absent', 'latest']),
+        version=dict(default=None, type='str'),
+    )
+    module = AnsibleModule(
+        argument_spec=arg_spec,
+        supports_check_mode=True
+    )
+
+    name = module.params['name']
+    path = module.params['path']
+    state = module.params['state']
+    version = module.params['version']
+
+    if not name:
+        module.fail_json(msg='Package name must be specified')
+
+    cargo = Cargo(module, name=name, path=path, state=state, version=version)
+
+    changed = False
+    if state == 'present':
+        # TODO write get_version function and installed_packages -> installed
+        if name not in cargo.installed_packages or version != cargo.installed_packages.get(name):
```

Comment:
This line causes the installed packages list to be retrieved twice in a row in some situations. You should probably cache its value locally.

Also it makes sense to convert `installed_packages` from an property to a function to make it clearer to the caller that using it is expensive.

Reply:
I was thinking of using memoization inside the property to make it cheaper on subsequent calls. But if you think it makes more sense to convert it to function, I can do that.

Reply:
If you use memoization, you have to take care that the cache is invalidated when you modify the state (like installing / uninstalling packages).

Reply:
You're right. Simple is better than complex :).